File: first_bad_version.ts
/**
 * The knows API is defined in the parent class Relation.
 * isBadVersion(version: number): boolean {
 *     ...
 * };
 */

function binarySearch(start: number, end: number, isBadVersion: any): number {
    let mid = Math.floor((start + end)/2)
    if (end - start <= 0) {
        return isBadVersion(start) ? start : -1
    }
    let mResult = isBadVersion(mid) ? mid : -1

    if (mResult > -1) {
        let lResult = binarySearch(start, mid-1, isBadVersion)
        return (lResult > -1) ? lResult : mResult
    } else {
        return binarySearch(mid+1, end, isBadVersion)
    }
}


var solution = function(isBadVersion: any) {

    return function(n: number): number {
        return binarySearch(1, n+1, isBadVersion)
    };
};

File: first_bad_version.swift
class Solution: VersionControl {

    func binarySearch(_ n: Int,_ left: Int,_ right: Int) -> Int {
        if left > right {
            return isBadVersion(left) ? left : -1
        }
        let pivot = left + (right - left) / 2
        let mResult = isBadVersion(pivot)
        if mResult == true {
            return binarySearch(n, left, pivot - 1)
        }
        return binarySearch(n, pivot + 1, right)
    }

    func firstBadVersion(_ n: Int) -> Int {
        binarySearch(n, 1, n)
    }
}
